### ISSUE TITLE
spatial attention trainable parameters

### ISSUE THREAD

**Issue by riyaj8888** [EscVM/EscVM_YT#2]
can you tell why there are zero trainable parameters in spatial attention layers?

in the notebook 3.3 build cnn with attention ,if we see the summary ,there are zero trainable parameters.

`Layer (type)                 Output Shape              Param #   
=================================================================
conv2d_6 (Conv2D)            (None, 32, 32, 32)        896       
_________________________________________________________________
batch_normalization_4 (Batch (None, 32, 32, 32)        128       
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 32, 32, 32)        9248      
_________________________________________________________________
batch_normalization_5 (Batch (None, 32, 32, 32)        128       
_________________________________________________________________
channel_attention (ChannelAt (None, 32, 32, 32)        0         
_________________________________________________________________
spatial_attention (SpatialAt (None, 32, 32, 32)        0         
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 16, 16, 32)        0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 16, 16, 64)        18496     
_________________________________________________________________
batch_normalization_6 (Batch (None, 16, 16, 64)        256       
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 16, 16, 64)        36928     
_________________________________________________________________
batch_normalization_7 (Batch (None, 16, 16, 64)        256       
_________________________________________________________________
channel_attention_1 (Channel (None, 16, 16, 64)        0         
_________________________________________________________________
spatial_attention_1 (Spatial (None, 16, 16, 64)        0         
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 8, 8, 64)          0         
_________________________________________________________________
conv2d_10 (Conv2D)           (None, 8, 8, 128)         73856     
_________________________________________________________________
batch_normalization_8 (Batch (None, 8, 8, 128)         512       
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 8, 8, 128)         147584    
_________________________________________________________________
channel_attention_2 (Channel (None, 8, 8, 128)         **0**         
_________________________________________________________________
spatial_attention_2 (Spatial (None, 8, 8, 128)         **0**         
_________________________________________________________________
global_average_pooling2d_1 ( (None, 128)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
=================================================================
Total params: 289,578
Trainable params: 288,938
Non-trainable params: 640
_________________________________________________________________`

**Comment by riyaj8888** [comment]
`Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv2d_54 (Conv2D)           (None, 32, 32, 32)        896       
_________________________________________________________________
batch_normalization_30 (Batc (None, 32, 32, 32)        128       
_________________________________________________________________
conv2d_55 (Conv2D)           (None, 32, 32, 32)        9248      
_________________________________________________________________
batch_normalization_31 (Batc (None, 32, 32, 32)        128       
_________________________________________________________________
channel_attention__18 (Chann (None, 32, 32, 32)        0         
_________________________________________________________________
spatial__attention__18 (Spat (None, 32, 32, 32)        98        
_________________________________________________________________
max_pooling2d_12 (MaxPooling (None, 16, 16, 32)        0         
_________________________________________________________________
conv2d_57 (Conv2D)           (None, 16, 16, 64)        18496     
_________________________________________________________________
batch_normalization_32 (Batc (None, 16, 16, 64)        256       
_________________________________________________________________
conv2d_58 (Conv2D)           (None, 16, 16, 64)        36928     
_________________________________________________________________
batch_normalization_33 (Batc (None, 16, 16, 64)        256       
_________________________________________________________________
channel_attention__19 (Chann (None, 16, 16, 64)        0         
_________________________________________________________________
spatial__attention__19 (Spat (None, 16, 16, 64)        98        
_________________________________________________________________
max_pooling2d_13 (MaxPooling (None, 8, 8, 64)          0         
_________________________________________________________________
conv2d_60 (Conv2D)           (None, 8, 8, 128)         73856     
_________________________________________________________________
batch_normalization_34 (Batc (None, 8, 8, 128)         512       
_________________________________________________________________
conv2d_61 (Conv2D)           (None, 8, 8, 128)         147584    
_________________________________________________________________
channel_attention__20 (Chann (None, 8, 8, 128)         0         
_________________________________________________________________
spatial__attention__20 (Spat (None, 8, 8, 128)         98        
_________________________________________________________________
global_average_pooling2d_6 ( (None, 128)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 10)                1290      
=================================================================
Total params: 289,872
Trainable params: 289,232
Non-trainable params: 640
_________________________________________________________________`

**Comment by riyaj8888** [comment]
`
class ChannelAttention_(tf.keras.layers.Layer):

  def __init__(self,factor):
    super(ChannelAttention_,self).__init__()
    self.factor = factor

  def call(self,x):

    bs,h,w,features = x.shape
    self.input_features = features

    x_avg = tf.keras.layers.GlobalAveragePooling2D()(x)

    x_avg = tf.keras.layers.Dense(self.input_features//self.factor,activation='relu',kernel_initializer='he_normal',use_bias=True, bias_initializer='zeros')(x_avg)
    x_avg = tf.keras.layers.Dense(self.input_features,kernel_initializer='he_normal',use_bias=True, bias_initializer='zeros')(x_avg)

    x_max = tf.keras.layers.GlobalMaxPooling2D()(x)
    x_max = tf.keras.layers.Dense(self.input_features//self.factor,activation='relu',kernel_initializer='he_normal',use_bias=True, bias_initializer='zeros')(x_max)
    x_max = tf.keras.layers.Dense(self.input_features,kernel_initializer='he_normal',use_bias=True, bias_initializer='zeros')(x_max)

    attention = tf.keras.layers.Add()([x_avg,x_max])
    attention = tf.keras.layers.Activation('sigmoid')(attention)

    channel_Attention_output = tf.keras.layers.Multiply()([x, attention])

    return channel_Attention_output
    
`

`class Spatial_Attention_(tf.keras.layers.Layer):

  def __init__(self,kernel_size):
    super(Spatial_Attention_,self).__init__()
    self.kernel_size=kernel_size
    self.conv2d = tf.keras.layers.Conv2D(filters = 1,
                  kernel_size=self.kernel_size,
                  strides=1,
                  padding='same',
                  activation='sigmoid',
                  kernel_initializer='he_normal',
                  use_bias=False)
              
  def call(self,x):

    avg_pool = tf.keras.layers.Lambda(lambda x: tf.keras.backend.mean(x, axis=3, keepdims=True))(x)
    max_pool = tf.keras.layers.Lambda(lambda x: tf.keras.backend.max(x, axis=3, keepdims=True))(x)
    attention = tf.keras.layers.Concatenate(axis=3)([avg_pool, max_pool])

    attention = self.conv2d(attention)

    spatial_Attention_output = tf.keras.layers.multiply([x, attention])

    return spatial_Attention_output
`
